### DISCUSSION TITLE
ESP32 maximum PWM Channels

### DISCUSSION THREAD

**Discussion by Gorgonbert** [arendst/Tasmota#10144] [Q&A] [answered]
Hi...

For a special project I would like to control the dimming of 6 white LED strips. I used PWM and MOSFETS on a QuinLED Deca board and it works great. Unfortunately I'm stuck with only 5 working strips because of the limitation of 5 PWM channels. 

I know this is a very special case and not many people use this many PWM channels, but may I kindly ask if this could be supported in the future? 

I tried to use ESPHome and there it works, so the Hardware should be fine. (I just like tasmota much better) 

The new GPIO numbering scheme has enough free numbers and would allow it. 

Please... pretty please ;-)

Cheers,

Mike

**Comment by s-hadinger** [comment]
I'm afraid the Light module is designed for 5 channels (RGBCW) and expanding this would require some work.

It may be easier to just extend the `PMW` command outside of lights, but changing `MAX_PWMS` from 5 to 6 would break the settings part and shift alignment.

**Comment by gemu2015** [comment]
the scripting language has a pwm() command which is currently limited to 3 channels.
with a simple patch you can expand this to 6  for the esp32 (it supports up to 16 pwm channels)

patch this line from 3 to 6
void esp_pwm(int32_t value, uint32 freq, uint32_t channel) {
  if (channel < 1 || channel > 3) channel = 1;

however you have to handle everything yourself and learn scripting.

**Comment by gemu2015** [comment, marked as answer]
i updated the pwm() cmd in scripter to support up to 8 PWM channels with scripts on ESP32

**Reply by l1t7l3ph0o7** [reply]
Can you please explain how you did this to a newbie? I would like 8 pwm channels for individual leds

**Reply by sfromis** [reply]
If you use the [Scripting-Language](https://tasmota.github.io/docs/Scripting-Language/) (takes some learning curve) to control the PWM leds, you have the support already.

**Reply by gemu2015** [reply]
make yourself familiar with scripting (see tasmota docs)

here is a script with 8 PWM channels for 8 LEDS to start with

\>D
led1=0
led2=0
led3=0
led4=0
led5=0
led6=0
led7=0
led8=0
on1=0
on2=0
on3=0
on4=0
on5=0
on6=0
on7=0
on8=0

\>B
; define pwm for pins 13 to 20 with 4000 Hz
pwm1(-13 4000)
pwm2(-14 4000)
pwm3(-15 4000)
pwm4(-16 4000)
pwm5(-17 4000)
pwm6(-18 4000)
pwm7(-19 4000)
pwm8(-20 4000)


\>F

if on1>0
then
	pwm1(led1)
else
	pwm1(0)
endif

if on2>0
then
	pwm2(led2)
else
	pwm2(0)
endif

if on3>0
then
	pwm3(led3)
else
	pwm3(0)
endif

if on4>0
then
	pwm4(led4)
else
	pwm4(0)
endif

if on5>0
then
	pwm5(led5)
else
	pwm5(0)
endif

if on6>0
then
	pwm6(led6)
else
	pwm6(0)
endif

if on7>0
then
	pwm7(led7)
else
	pwm7(0)
endif

if on8>0
then
	pwm8(led8)
else
	pwm8(0)
endif



\>W
bu(on1 "on" "off")
sl(0 1023 led1 "0" "LED1" "100")
bu(on2 "on" "off")
sl(0 1023 led2 "0" "LED2" "100")
bu(on3 "on" "off")
sl(0 1023 led3 "0" "LED3" "100")
bu(on4 "on" "off")
sl(0 1023 led4 "0" "LED4" "100")
bu(on5 "on" "off")
sl(0 1023 led5 "0" "LED5" "100")
bu(on6 "on" "off")
sl(0 1023 led6 "0" "LED6" "100")
bu(on7 "on" "off")
sl(0 1023 led7 "0" "LED7" "100")
bu(on8 "on" "off")
sl(0 1023 led8 "0" "LED8" "100")

**Reply by l1t7l3ph0o7** [reply]
Thank you so much.  One more thing. Can you show an example of how to control 1 ch via mqtt? I’m sorry if its painfully obvious, but I’m struggling to understand.

**Reply by l1t7l3ph0o7** [reply]
I get this error when I copy and paste your script into Berry scripting console: input:1: unexpected symbol near '>' should I be creating a file in "Manage file system"? if so with what extension?

**Reply by barbudor** [reply]
This is not Berry, this is [Scripting-Language](https://tasmota.github.io/docs/Scripting-Language)

**Comment by dr-apple** [comment]
Hello please added support for all 16 pwm channel with fade ;-)
